File: scripts/render_tiktok_slideshows.swift
import AppKit
import Foundation

let width: CGFloat = 1080
let height: CGFloat = 1920
let outputURL = URL(fileURLWithPath: "Marketing/TikTok/slideshows")

struct Theme {
    let bg1: NSColor
    let bg2: NSColor
    let bg3: NSColor
    let ink: NSColor
    let chip: NSColor
    let soft: NSColor
    let line: NSColor
}

struct Deck {
    let id: String
    let title: String
    let kicker: String
    let caption: String
    let comment: String
    let theme: String
    let style: String
    let slides: [String]
}

func color(_ hex: String) -> NSColor {
    let clean = hex.trimmingCharacters(in: CharacterSet(charactersIn: "#"))
    var value: UInt64 = 0
    Scanner(string: clean).scanHexInt64(&value)
    return NSColor(
        calibratedRed: CGFloat((value >> 16) & 0xff) / 255,
        green: CGFloat((value >> 8) & 0xff) / 255,
        blue: CGFloat(value & 0xff) / 255,
        alpha: 1
    )
}

let themes: [String: Theme] = [
    "sun": Theme(bg1: color("#f6f0df"), bg2: color("#c7a66a"), bg3: color("#183b35"), ink: color("#14251f"), chip: color("#17352f"), soft: color("#f9f4e8"), line: color("#c3a35e")),
    "pool": Theme(bg1: color("#eef4ed"), bg2: color("#a9c7bf"), bg3: color("#173a4a"), ink: color("#10242c"), chip: color("#123544"), soft: color("#f4f0e6"), line: color("#b89555")),
    "gold": Theme(bg1: color("#f7f0df"), bg2: color("#d1b36f"), bg3: color("#26352c"), ink: color("#1a241d"), chip: color("#223027"), soft: color("#fbf5e6"), line: color("#b99453")),
    "travel": Theme(bg1: color("#f4ead6"), bg2: color("#b98f58"), bg3: color("#1c2f4a"), ink: color("#182235"), chip: color("#182a42"), soft: color("#f8f1e5"), line: color("#b89555"))
]

let decks: [Deck] = [
    Deck(
        id: "01-mare-tan-vibe",
        title: "Riviera tan routine",
        kicker: "RIVIERA NOTES",
        caption: "Routine mare: luce giusta, lato giusto, diario tan. Niente caos.",
        comment: "Commenta RIVIERA + fototipo e preparo una routine.",
        theme: "sun",
        style: "card",
        slides: [
            "routine mare da fare con calma",
            "linen shirt, acqua fredda, stessa luce",
            "il punto e' non andare a memoria",
            "a meta' apro Solea",
            "timer fronte e retro",
            "diario tan a fine giornata",
            "commenta RIVIERA + fototipo"
        ]
    ),
    Deck(
        id: "02-piscina-tan-routine",
        title: "Pool club routine",
        kicker: "POOL CLUB",
        caption: "Lettino, tuffo, lato: routine piscina con ritmo, non con fretta.",
        comment: "Team mare o piscina?",
        theme: "pool",
        style: "card",
        slides: [
            "pool club tan routine",
            "lettino, tuffo, lato A",
            "il ritmo fa la differenza",
            "qui uso Solea come tan timer",
            "durata, lato, pause, diario",
            "domani replico quello che funziona",
            "team mare o piscina?"
        ]
    ),
    Deck(
        id: "03-abbronzatura-uniforme",
        title: "Uniform tan notes",
        kicker: "UNIFORM TAN",
        caption: "Il tan uniforme non e' fortuna: lato, zone, costanza.",
        comment: "Quale zona ti si abbronza sempre peggio?",
        theme: "gold",
        style: "card",
        slides: [
            "abbronzatura uniforme",
            "non e' fortuna",
            "e' lato, zone, costanza",
            "qui entra Solea",
            "fronte e retro separati",
            "zone esposte nel diario",
            "meno caso, piu' risultato"
        ]
    ),
    Deck(
        id: "04-base-vacanza",
        title: "Resort base tan",
        kicker: "RESORT NOTES",
        caption: "Se parti tra pochi giorni, serve una base graduale, non fretta.",
        comment: "Dove parti? Scrivi meta + giorni alla partenza.",
        theme: "travel",
        style: "card",
        slides: [
            "parto tra 7 giorni",
            "voglio una base pulita",
            "non tutto il primo giorno",
            "a meta' apro Solea",
            "meta, date, UV, fototipo",
            "piano leggero + diario tan",
            "commenta la meta"
        ]
    ),
    Deck(
        id: "05-routine-mare-aesthetic",
        title: "Riviera aesthetic",
        kicker: "RIVIERA ROUTINE",
        caption: "Routine mare old money: pochi gesti, stesso ritmo, Solea solo quando serve.",
        comment: "Vuoi versione piscina o barca?",
        theme: "sun",
        style: "aesthetic",
        slides: [
            "routine mare old money",
            "lino, acqua fredda, lato A",
            "niente fretta, stesso ritmo",
            "a meta' apro Solea",
            "timer fronte / retro",
            "diario tan a fine giornata",
            "salva questa routine"
        ]
    ),
    Deck(
        id: "06-routine-piscina-aesthetic",
        title: "Poolside aesthetic",
        kicker: "POOLSIDE ROUTINE",
        caption: "Poolside tan: lettino, lato, pausa, diario. Product placement velato.",
        comment: "Team mare o piscina?",
        theme: "pool",
        style: "aesthetic",
        slides: [
            "poolside tan routine",
            "lettino, tuffo, lato A",
            "il ritmo fa il risultato",
            "qui entra Solea",
            "timer tan + pause in acqua",
            "domani replico quello che funziona",
            "team mare o piscina?"
        ]
    ),
    Deck(
        id: "07-tan-uniforme-aesthetic",
        title: "Tennis club tan",
        kicker: "TENNIS CLUB TAN",
        caption: "Uniform tan, ma con mood tennis club: lato, zone, diario.",
        comment: "Quale zona ti si abbronza sempre peggio?",
        theme: "gold",
        style: "aesthetic",
        slides: [
            "uniform tan notes",
            "girarsi prima di ricordarselo",
            "seguire le zone, non solo il tempo",
            "a meta' routine uso Solea",
            "fronte, retro, zone esposte",
            "ripeto quello che funziona",
            "quale punto ti manca?"
        ]
    )
]

func topRect(_ x: CGFloat, _ y: CGFloat, _ w: CGFloat, _ h: CGFloat) -> NSRect {
    NSRect(x: x, y: height - y - h, width: w, height: h)
}

func rounded(_ rect: NSRect, _ radius: CGFloat, fill: NSColor, alpha: CGFloat = 1) {
    fill.withAlphaComponent(alpha).setFill()
    NSBezierPath(roundedRect: rect, xRadius: radius, yRadius: radius).fill()
}

func strokeRounded(_ rect: NSRect, _ radius: CGFloat, color: NSColor, alpha: CGFloat = 1, lineWidth: CGFloat = 2) {
    let path = NSBezierPath(roundedRect: rect, xRadius: radius, yRadius: radius)
    path.lineWidth = lineWidth
    color.withAlphaComponent(alpha).setStroke()
    path.stroke()
}

func serifFont(size: CGFloat) -> NSFont {
    NSFont(name: "Didot", size: size)
        ?? NSFont(name: "Bodoni 72 Book", size: size)
        ?? NSFont(name: "Georgia", size: size)
        ?? NSFont.systemFont(ofSize: size, weight: .regular)
}

func sansFont(size: CGFloat, weight: NSFont.Weight) -> NSFont {
    NSFont.systemFont(ofSize: size, weight: weight)
}

func drawText(_ text: String, in rect: NSRect, size: CGFloat, weight: NSFont.Weight, color: NSColor, alignment: NSTextAlignment = .left, serif: Bool = false, kern: CGFloat = 0) {
    let paragraph = NSMutableParagraphStyle()
    paragraph.alignment = alignment
    paragraph.lineBreakMode = .byWordWrapping
    let attrs: [NSAttributedString.Key: Any] = [
        .font: serif ? serifFont(size: size) : sansFont(size: size, weight: weight),
        .foregroundColor: color,
        .paragraphStyle: paragraph,
        .kern: kern
    ]
    (text as NSString).draw(with: rect, options: [.usesLineFragmentOrigin, .usesFontLeading], attributes: attrs)
}

func drawCentered(_ text: String, centerY: CGFloat, size: CGFloat, color: NSColor) {
    let rect = topRect(100, centerY - 250, 880, 500)
    let paragraph = NSMutableParagraphStyle()
    paragraph.alignment = .center
    let attrs: [NSAttributedString.Key: Any] = [
        .font: serifFont(size: size),
        .foregroundColor: color,
        .paragraphStyle: paragraph,
        .kern: -0.5
    ]
    let bounds = (text as NSString).boundingRect(
        with: NSSize(width: rect.width, height: 900),
        options: [.usesLineFragmentOrigin, .usesFontLeading],
        attributes: attrs
    )
    let drawRect = NSRect(x: rect.minX, y: rect.midY - bounds.height / 2, width: rect.width, height: bounds.height + 20)
    (text as NSString).draw(with: drawRect, options: [.usesLineFragmentOrigin, .usesFontLeading], attributes: attrs)
}

func drawBackground(theme: Theme) {
    NSGradient(colors: [theme.bg1, theme.bg2, theme.bg3])?.draw(in: topRect(0, 0, width, height), angle: -42)
    theme.soft.withAlphaComponent(0.22).setFill()
    topRect(72, 246, 936, 1220).fill()
    strokeRounded(topRect(72, 246, 936, 1220), 0, color: theme.line, alpha: 0.32, lineWidth: 2)
    theme.line.withAlphaComponent(0.18).setStroke()
    let line = NSBezierPath()
    line.move(to: NSPoint(x: 135, y: height - 320))
    line.line(to: NSPoint(x: 945, y: height - 320))
    line.lineWidth = 2
    line.stroke()
}

func backgroundImageURL(deck: Deck, index: Int) -> URL? {
    let base = URL(fileURLWithPath: "Marketing/TikTok/backgrounds")
        .appendingPathComponent(deck.id)
        .appendingPathComponent(String(format: "%02d", index + 1))
    for ext in ["jpg", "jpeg", "png"] {
        let candidate = base.appendingPathExtension(ext)
        if FileManager.default.fileExists(atPath: candidate.path) {
            return candidate
        }
    }
    return nil
}

func drawPhotoBackgroundIfAvailable(deck: Deck, index: Int, theme: Theme) -> Bool {
    guard
        let url = backgroundImageURL(deck: deck, index: index),
        let image = NSImage(contentsOf: url),
        let best = image.representations.first
    else {
        return false
    }

    let imageSize = NSSize(width: best.pixelsWide, height: best.pixelsHigh)
    let scale = max(width / imageSize.width, height / imageSize.height)
    let drawSize = NSSize(width: imageSize.width * scale, height: imageSize.height * scale)
    let rect = NSRect(
        x: (width - drawSize.width) / 2,
        y: (height - drawSize.height) / 2,
        width: drawSize.width,
        height: drawSize.height
    )
    image.draw(in: rect, from: .zero, operation: .sourceOver, fraction: 1)
    NSColor.black.withAlphaComponent(0.16).setFill()
    topRect(0, 0, width, height).fill()
    return true
}

func drawTopBar(deck: Deck, index: Int, total: Int, theme: Theme, branded: Bool) {
    rounded(topRect(72, 82, 936, 86), 8, fill: theme.chip, alpha: branded ? 0.76 : 0.62)
    strokeRounded(topRect(72, 82, 936, 86), 8, color: theme.line, alpha: 0.70, lineWidth: 1.5)
    let left = branded ? "SOLEA" : "RIVIERA"
    let mid = branded ? "TAN COMPANION" : deck.kicker
    drawText(left, in: topRect(116, 108, 220, 44), size: 27, weight: .bold, color: .white, serif: true, kern: 1.2)
    drawText(mid, in: topRect(350, 112, 430, 44), size: 20, weight: .semibold, color: NSColor.white.withAlphaComponent(0.78), kern: 2.8)
    drawText(String(format: "%02d/%02d", index + 1, total), in: topRect(848, 111, 120, 44), size: 22, weight: .medium, color: NSColor.white.withAlphaComponent(0.82), alignment: .right, kern: 1.2)
}

func drawChips(theme: Theme, branded: Bool) {
    let values = branded
        ? [("UV", "reale"), ("FOTOTIPO", "tuo"), ("LATO", "on"), ("DIARIO", "save")]
        : [("VIBE", "mare"), ("TAN", "plan"), ("LATO", "mix"), ("MOOD", "estate")]
    for (idx, item) in values.enumerated() {
        let x = CGFloat(120 + idx * 210)
        rounded(topRect(x, 1390, 178, 154), 8, fill: theme.soft, alpha: 0.72)
        strokeRounded(topRect(x, 1390, 178, 154), 8, color: theme.line, alpha: 0.36, lineWidth: 1)
        drawText(item.0, in: topRect(x + 22, 1424, 134, 32), size: 18, weight: .semibold, color: theme.ink.withAlphaComponent(0.56), kern: 1.8)
        drawText(item.1, in: topRect(x + 22, 1470, 140, 54), size: item.1.count > 5 ? 30 : 38, weight: .medium, color: theme.ink, serif: true)
    }
}

func drawProductCard(theme: Theme) {
    rounded(topRect(230, 1048, 620, 216), 10, fill: theme.soft, alpha: 0.86)
    strokeRounded(topRect(230, 1048, 620, 216), 10, color: theme.line, alpha: 0.60, lineWidth: 1.5)
    drawText("SOLEA", in: topRect(274, 1088, 220, 42), size: 25, weight: .medium, color: theme.ink.withAlphaComponent(0.72), serif: true, kern: 1.8)
    drawText("tan companion", in: topRect(274, 1138, 460, 70), size: 46, weight: .regular, color: theme.ink, serif: true)
    drawText("front / back / diary", in: topRect(274, 1210, 500, 42), size: 23, weight: .medium, color: theme.ink.withAlphaComponent(0.62), kern: 1.2)
}

func renderCard(deck: Deck, slide: String, index: Int, total: Int, theme: Theme) {
    let branded = index >= max(3, total / 2)
    if !drawPhotoBackgroundIfAvailable(deck: deck, index: index, theme: theme) {
        drawBackground(theme: theme)
    }
    drawTopBar(deck: deck, index: index, total: total, theme: theme, branded: branded)
    rounded(topRect(92, 330, 896, 870), 10, fill: theme.soft, alpha: 0.78)
    strokeRounded(topRect(92, 330, 896, 870), 10, color: theme.line, alpha: 0.36, lineWidth: 1.5)
    let label = branded ? "SOLEA" : "NOTES"
    drawText(label, in: topRect(142, 386, 240, 40), size: 20, weight: .semibold, color: theme.ink.withAlphaComponent(0.54), kern: 2.4)
    drawText(slide, in: topRect(142, 502, 790, 560), size: slide.count > 58 ? 66 : 82, weight: .regular, color: theme.ink, serif: true)
    drawChips(theme: theme, branded: branded)
    rounded(topRect(80, 1698, 920, 116), 8, fill: theme.chip, alpha: 0.86)
    strokeRounded(topRect(80, 1698, 920, 116), 8, color: theme.line, alpha: 0.55, lineWidth: 1)
    drawText(branded ? "SOLEA, TAN COMPANION" : "SLOW TAN, BETTER NOTES", in: topRect(124, 1728, 830, 34), size: 20, weight: .semibold, color: .white.withAlphaComponent(0.78), kern: 2.2)
    drawText(branded ? "UV, fototipo, lato, diario." : "Abbronzatura con metodo, non a caso.", in: topRect(124, 1764, 830, 38), size: 24, weight: .regular, color: .white.withAlphaComponent(0.76), serif: true)
}

func renderAesthetic(deck: Deck, slide: String, index: Int, total: Int, theme: Theme) {
    let branded = index >= max(3, total / 2)
    if !drawPhotoBackgroundIfAvailable(deck: deck, index: index, theme: theme) {
        drawBackground(theme: theme)
    }
    NSColor.black.withAlphaComponent(branded ? 0.18 : 0.10).setFill()
    topRect(0, 0, width, height).fill()
    drawTopBar(deck: deck, index: index, total: total, theme: theme, branded: branded)
    drawCentered(slide, centerY: 735, size: slide.count > 42 ? 68 : 80, color: .white)
    if branded {
        drawProductCard(theme: theme)
    }
    rounded(topRect(72, 1698, 936, 104), 8, fill: theme.chip, alpha: 0.62)
    strokeRounded(topRect(72, 1698, 936, 104), 8, color: theme.line, alpha: 0.46, lineWidth: 1)
    drawText(branded ? "Solea entra solo quando serve: timer, lato, diario." : "linen, water, light, rhythm", in: topRect(116, 1733, 850, 42), size: 23, weight: .medium, color: .white.withAlphaComponent(0.88), serif: true)
}

func renderPNG(deck: Deck, slide: String, index: Int, total: Int, to url: URL) throws {
    guard let rep = NSBitmapImageRep(
        bitmapDataPlanes: nil,
        pixelsWide: Int(width),
        pixelsHigh: Int(height),
        bitsPerSample: 8,
        samplesPerPixel: 4,
        hasAlpha: true,
        isPlanar: false,
        colorSpaceName: .deviceRGB,
        bytesPerRow: 0,
        bitsPerPixel: 0
    ) else {
        throw NSError(domain: "SoleaTikTok", code: 1)
    }
    NSGraphicsContext.saveGraphicsState()
    NSGraphicsContext.current = NSGraphicsContext(bitmapImageRep: rep)
    let theme = themes[deck.theme]!
    if deck.style == "aesthetic" {
        renderAesthetic(deck: deck, slide: slide, index: index, total: total, theme: theme)
    } else {
        renderCard(deck: deck, slide: slide, index: index, total: total, theme: theme)
    }
    NSGraphicsContext.restoreGraphicsState()
    guard let data = rep.representation(using: .png, properties: [:]) else {
        throw NSError(domain: "SoleaTikTok", code: 2)
    }
    try data.write(to: url)
}

let fm = FileManager.default
if fm.fileExists(atPath: outputURL.path) {
    try fm.removeItem(at: outputURL)
}
try fm.createDirectory(at: outputURL, withIntermediateDirectories: true)

var readme: [String] = [
    "# Solea TikTok slideshow assets",
    "",
    "Direzione: old money/resort, marketing velato. Le prime slide sono vibe/utility tan-first; Solea entra a meta' carousel.",
    "",
    "Carica i PNG in ordine numerico dentro ogni cartella.",
    ""
]
var pngCount = 0

for deck in decks {
    let deckURL = outputURL.appendingPathComponent(deck.id)
    try fm.createDirectory(at: deckURL, withIntermediateDirectories: true)
    readme.append("## \(deck.title)")
    readme.append("")
    readme.append("Directory: `\(deckURL.path)`")
    readme.append("Caption: \(deck.caption)")
    readme.append("Commento fissato: \(deck.comment)")
    readme.append("")
    for (idx, slide) in deck.slides.enumerated() {
        let fileURL = deckURL.appendingPathComponent(String(format: "%02d.png", idx + 1))
        try renderPNG(deck: deck, slide: slide, index: idx, total: deck.slides.count, to: fileURL)
        readme.append("\(idx + 1). \(slide)")
        pngCount += 1
    }
    readme.append("")
}

try readme.joined(separator: "\n").write(to: outputURL.appendingPathComponent("README.md"), atomically: true, encoding: .utf8)
print("Rendered \(decks.count) slideshows to \(outputURL.path)")
print("PNG files generated: \(pngCount)")

File: scripts/render-tiktok-slideshows.mjs
#!/usr/bin/env node

import { existsSync, mkdirSync, renameSync, rmSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { spawnSync } from "node:child_process";

const WIDTH = 1080;
const HEIGHT = 1920;
const OUT_DIR = "Marketing/TikTok/slideshows";

const slideshows = [
  {
    id: "01-primo-giorno-di-mare",
    title: "Primo giorno di mare",
    caption:
      "Il tan dell'estate si costruisce il primo giorno, non l'ultimo.",
    pinnedComment: "Fototipo + citta' e ti dico il piano tan di oggi.",
    theme: "sun",
    slides: [
      "Il primo giorno di mare decide tutta la tua estate.",
      "Se lo sbagli, ti sbucci e riparti da zero.",
      "Io lo imposto in 30 secondi: UV reale + fototipo + sole gia' preso.",
      "Tanora lo trasforma in timer: quanto, quando, da che lato.",
      "Colore che resta. Niente chiazze, niente sbucciature.",
      "Commenta MARE + fototipo e ti mostro il piano di oggi."
    ]
  },
  {
    id: "02-abbronzatura-a-chiazze",
    title: "Tan uniforme",
    caption:
      "Tan uniforme non e' fortuna: e' sapere quando girarti.",
    pinnedComment: "Quale zona ti resta sempre piu' chiara?",
    theme: "gold",
    slides: [
      "Il motivo per cui ti abbronzi a chiazze.",
      "Non e' il sole. E' che non tieni il conto dei lati.",
      "Io uso un timer fronte/retro: mi dice quando girarmi.",
      "Segno le zone esposte: viso, spalle, gambe, schiena.",
      "Risultato: colore uniforme, non a strisce.",
      "Salva per la prossima giornata al sole."
    ]
  },
  {
    id: "03-golden-tan-hours",
    title: "Golden tan hours",
    caption:
      "Il tan e' una questione di timing, non di ore in piu'.",
    pinnedComment: "Citta' + fototipo e ti dico le tue golden tan hours.",
    theme: "sun",
    slides: [
      "L'orario in cui ti abbronzi meglio non e' quello che pensi.",
      "Il sole delle 13 ti cuoce. Quello giusto ti colora.",
      "Tanora legge la curva UV della tua giornata.",
      "E mi da' le golden tan hours per il mio fototipo.",
      "Stesso colore, zero sbucciature: solo timing.",
      "Commenta citta' + fototipo e ti dico le tue."
    ]
  },
  {
    id: "04-base-tan-vacanza",
    title: "Base tan",
    caption:
      "La base tan si costruisce prima di partire.",
    pinnedComment: "Meta + giorni alla partenza e ti faccio il piano base tan.",
    theme: "travel",
    slides: [
      "Base tan in 7 giorni prima della partenza.",
      "Cosi' il primo giorno di vacanza parti gia' con il colore.",
      "Tanora parte da meta, date e UV dei prossimi giorni.",
      "E costruisce il piano giorno per giorno.",
      "Arrivo in vacanza con la base, non con la fretta.",
      "Dove parti? Scrivi meta + giorni alla partenza."
    ]
  },
  {
    id: "05-piscina-vs-mare",
    title: "Piscina vs mare",
    caption:
      "La piscina e' sottovalutata per il tan: basta tenere il conto.",
    pinnedComment: "Team mare o team piscina?",
    theme: "pool",
    slides: [
      "Piscina o mare per abbronzarsi? Facciamo i conti.",
      "In piscina il riflesso dell'acqua lavora per te.",
      "Ma tra un tuffo e l'altro perdi il conto. Sempre.",
      "Timer: pausa quando entro in acqua, riprendo al lettino.",
      "Fine giornata: tutto nel diario tan, colore tracciato.",
      "Team mare o team piscina? Commenta."
    ]
  },
  {
    id: "06-diario-tan-30-giorni",
    title: "Diario tan",
    caption:
      "Il before/after funziona solo se lo fai con metodo.",
    pinnedComment: "Lo faresti un diario tan di 30 giorni?",
    theme: "gold",
    slides: [
      "Ho tracciato la mia abbronzatura per 30 giorni: risultati.",
      "Prima/dopo vero: stessa luce, stesso angolo, stesso orario.",
      "Le foto restano sul mio telefono, non su un server.",
      "Tanora mi mostra il progresso settimana per settimana.",
      "Vedere il colore salire e' la motivazione migliore.",
      "Salva se vuoi iniziare il tuo diario tan."
    ]
  }
];

const themes = {
  sun: {
    bg1: "#fff2bc",
    bg2: "#ffbf55",
    bg3: "#f36a32",
    ink: "#17110b",
    accent: "#ffffff",
    chip: "#20130c",
    soft: "#fff8dd"
  },
  check: {
    bg1: "#fff7df",
    bg2: "#ffd36e",
    bg3: "#ef7a3a",
    ink: "#151515",
    accent: "#ffffff",
    chip: "#171717",
    soft: "#fffaf0"
  },
  gold: {
    bg1: "#fff6d8",
    bg2: "#e9b86a",
    bg3: "#c9773d",
    ink: "#17110b",
    accent: "#ffffff",
    chip: "#2a190e",
    soft: "#fff6df"
  },
  pool: {
    bg1: "#e8fbff",
    bg2: "#72d7ef",
    bg3: "#ffb45f",
    ink: "#101820",
    accent: "#ffffff",
    chip: "#10242c",
    soft: "#f2fdff"
  },
  travel: {
    bg1: "#fff1c7",
    bg2: "#ff9f6a",
    bg3: "#e8573d",
    ink: "#1a1110",
    accent: "#ffffff",
    chip: "#2b1713",
    soft: "#fff0d8"
  }
};

function escapeXml(value) {
  return String(value)
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;");
}

function ensureDir(path) {
  if (!existsSync(path)) {
    mkdirSync(path, { recursive: true });
  }
}

function wrapWords(text, maxChars) {
  const words = text.split(/\s+/);
  const lines = [];
  let current = "";

  for (const word of words) {
    const next = current ? `${current} ${word}` : word;
    if (next.length > maxChars && current) {
      lines.push(current);
      current = word;
    } else {
      current = next;
    }
  }

  if (current) {
    lines.push(current);
  }
  return lines;
}

function textBlock({
  text,
  x,
  y,
  width,
  size,
  weight = 800,
  fill,
  anchor = "start",
  lineHeight = 1.05,
  maxLines = 7
}) {
  const avgCharWidth = size * 0.54;
  const maxChars = Math.max(10, Math.floor(width / avgCharWidth));
  const lines = wrapWords(text, maxChars).slice(0, maxLines);
  return lines
    .map((line, index) => {
      const lineY = y + index * size * lineHeight;
      return `<text x="${x}" y="${lineY}" text-anchor="${anchor}" font-family="Inter, SF Pro Display, Arial, sans-serif" font-size="${size}" font-weight="${weight}" fill="${fill}">${escapeXml(line)}</text>`;
    })
    .join("");
}

function metricsCard(theme, slideIndex) {
  const metrics = [
    ["UV", slideIndex < 2 ? "oggi" : "reale"],
    ["FOTOTIPO", "tuo"],
    ["LATO", "on"],
    ["DIARIO", "save"]
  ];

  const items = metrics
    .map(([label, value], index) => {
      const x = 120 + index * 210;
      return `
        <g transform="translate(${x} 0)">
          <rect width="178" height="154" rx="32" fill="${theme.accent}" opacity="0.58"/>
          <text x="24" y="50" font-family="Inter, SF Pro Display, Arial, sans-serif" font-size="26" font-weight="900" fill="${theme.ink}" opacity="0.55">${label}</text>
          <text x="24" y="110" font-family="Inter, SF Pro Display, Arial, sans-serif" font-size="46" font-weight="950" fill="${theme.ink}">${value}</text>
        </g>`;
    })
    .join("");

  return `
    <g transform="translate(0 1390)">
      ${items}
    </g>`;
}

function iconFor(slideText) {
  const lower = slideText.toLowerCase();
  if (lower.includes("golden")) return "GOLDEN";
  if (lower.includes("diario") || lower.includes("foto")) return "DIARIO";
  if (lower.includes("lato") || lower.includes("fronte") || lower.includes("retro")) return "LATO";
  if (lower.includes("vacanza") || lower.includes("parti") || lower.includes("meta")) return "TRIP";
  if (lower.includes("timer")) return "TIMER";
  if (lower.includes("commenta")) return "CTA";
  if (lower.includes("salva")) return "SAVE";
  if (lower.includes("uv")) return "UV";
  if (lower.includes("fototipo")) return "SKIN";
  if (lower.includes("piscina")) return "POOL";
  if (lower.includes("mare")) return "SEA";
  return "CHECK";
}

function slideSvg({ slideshow, slideText, slideIndex, total }) {
  const theme = themes[slideshow.theme];
  const icon = iconFor(slideText);
  const slideNo = `${String(slideIndex + 1).padStart(2, "0")}/${String(total).padStart(2, "0")}`;
  const isOpening = slideIndex === 0;
  const isClosing = slideIndex === total - 1;
  const headlineSize = isOpening ? 116 : slideText.length > 58 ? 88 : 102;
  const kicker = isClosing ? "TOCCA A TE" : slideshow.title.toUpperCase();

  // Regola tan-first: la prima slide e' solo hook, niente brand.
  const header = isOpening
    ? ""
    : `<g transform="translate(80 92)">
    <rect width="920" height="96" rx="48" fill="${theme.chip}" opacity="0.94"/>
    <text x="48" y="62" font-family="Inter, SF Pro Display, Arial, sans-serif" font-size="34" font-weight="950" letter-spacing="4" fill="#ffffff">TANORA</text>
    <text x="250" y="62" font-family="Inter, SF Pro Display, Arial, sans-serif" font-size="27" font-weight="850" letter-spacing="2" fill="#ffffff" opacity="0.78">${escapeXml(kicker)}</text>
    <text x="858" y="62" text-anchor="end" font-family="Inter, SF Pro Display, Arial, sans-serif" font-size="30" font-weight="900" fill="#ffffff" opacity="0.86">${slideNo}</text>
  </g>`;

  return `<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" width="${WIDTH}" height="${HEIGHT}" viewBox="0 0 ${WIDTH} ${HEIGHT}">
  <defs>
    <linearGradient id="bg" x1="0" x2="1" y1="0" y2="1">
      <stop offset="0%" stop-color="${theme.bg1}"/>
      <stop offset="52%" stop-color="${theme.bg2}"/>
      <stop offset="100%" stop-color="${theme.bg3}"/>
    </linearGradient>
    <filter id="shadow" x="-20%" y="-20%" width="140%" height="140%">
      <feDropShadow dx="0" dy="24" stdDeviation="20" flood-color="#000000" flood-opacity="0.18"/>
    </filter>
  </defs>

  <rect width="${WIDTH}" height="${HEIGHT}" fill="url(#bg)"/>
  <path d="M0 1440 C220 1340 430 1450 650 1360 C835 1285 945 1265 1080 1320 L1080 1920 L0 1920 Z" fill="${theme.soft}" opacity="0.82"/>
  <path d="M-40 250 C230 140 520 210 780 120 C910 75 1010 80 1130 130" fill="none" stroke="${theme.accent}" stroke-width="18" opacity="0.34"/>
  <circle cx="878" cy="282" r="118" fill="${theme.accent}" opacity="0.24"/>

  ${header}

  <g transform="translate(80 308)">
    <rect width="920" height="920" rx="56" fill="${theme.accent}" opacity="0.70" filter="url(#shadow)"/>
    <text x="60" y="100" font-family="Inter, SF Pro Display, Arial, sans-serif" font-size="34" font-weight="950" letter-spacing="3" fill="${theme.ink}" opacity="0.54">${icon}</text>
    ${textBlock({
      text: slideText,
      x: 60,
      y: 270,
      width: 790,
      size: headlineSize,
      weight: 950,
      fill: theme.ink,
      lineHeight: 1.0,
      maxLines: 6
    })}
  </g>

  ${metricsCard(theme, slideIndex)}

  <g transform="translate(80 1698)">
    <rect width="920" height="116" rx="36" fill="${theme.chip}" opacity="0.94"/>
    <text x="44" y="48" font-family="Inter, SF Pro Display, Arial, sans-serif" font-size="26" font-weight="900" fill="#ffffff" opacity="0.78">ABBRONZATURA, MA CON METODO</text>
    <text x="44" y="86" font-family="Inter, SF Pro Display, Arial, sans-serif" font-size="24" font-weight="750" fill="#ffffff" opacity="0.72">UV, fototipo, lato, diario.</text>
  </g>
</svg>`;
}

function convertSvgToPng(svgPath, pngPath) {
  const sipsResult = spawnSync("sips", ["-s", "format", "png", svgPath, "--out", pngPath], {
    encoding: "utf8"
  });

  if (sipsResult.status === 0 && existsSync(pngPath)) {
    return true;
  }

  const tmpDir = join(dirname(pngPath), ".quicklook");
  ensureDir(tmpDir);
  const qlResult = spawnSync("qlmanage", ["-t", "-s", String(HEIGHT), "-o", tmpDir, svgPath], {
    encoding: "utf8"
  });
  const qlOutput = join(tmpDir, `${svgPath.split("/").pop()}.png`);
  if (qlResult.status === 0 && existsSync(qlOutput)) {
    renameSync(qlOutput, pngPath);
    return true;
  }

  return false;
}

function render() {
  if (existsSync(OUT_DIR)) {
    rmSync(OUT_DIR, { recursive: true, force: true });
  }
  ensureDir(OUT_DIR);
  const manifest = [];
  let pngCount = 0;

  for (const slideshow of slideshows) {
    const dir = join(OUT_DIR, slideshow.id);
    ensureDir(dir);
    manifest.push(`# ${slideshow.title}`);
    manifest.push("");
    manifest.push(`Directory: \`${dir}\``);
    manifest.push("");
    manifest.push(`Caption: ${slideshow.caption}`);
    manifest.push(`Commento fissato: ${slideshow.pinnedComment}`);
    manifest.push("");

    slideshow.slides.forEach((slideText, index) => {
      const basename = `${String(index + 1).padStart(2, "0")}.svg`;
      const svgPath = join(dir, basename);
      const pngPath = join(dir, `${String(index + 1).padStart(2, "0")}.png`);
      writeFileSync(
        svgPath,
        slideSvg({
          slideshow,
          slideText,
          slideIndex: index,
          total: slideshow.slides.length
        }),
        "utf8"
      );
      if (convertSvgToPng(svgPath, pngPath)) {
        pngCount += 1;
      }
      manifest.push(`${index + 1}. ${slideText}`);
    });
    manifest.push("");
  }

  writeFileSync(join(OUT_DIR, "MANIFEST.md"), `${manifest.join("\n")}\n`, "utf8");
  console.log(`Rendered ${slideshows.length} slideshows to ${OUT_DIR}`);
  console.log(`PNG files generated: ${pngCount}`);
}

render();
